src/main/scala/io/vlach/CheckoutSystem.scala

```diff
@@ -5,34 +5,17 @@ case object Apple extends Fruit(cost = 60L)
 case object Orange extends Fruit(cost = 25L)
 
 object CheckoutSystem {
-  def checkout(items: List[Fruit]): Long = {
-    val (applesUntyped, orangesUntyped) = items.partition {
-      case Apple => true
-      case Orange => false
-    }
-    val apples = applesUntyped.collect { case Apple => Apple }
-    val oranges = orangesUntyped.collect { case Orange => Orange }
-
-    val applesToCheckout = twoForOne(apples)
-    val orangesToCheckout = threeForTwo(oranges)
 
-    (applesToCheckout ++ orangesToCheckout).foldRight(0L)((fruit, acc) => fruit.cost + acc)
+  def checkout(items: List[Fruit]): Long = {
+    val (apples, oranges) = items.partition(_ == Apple)
+    (twoForOne(apples) ++ threeForTwo(oranges)).map(_.cost).sum
   }
 
-  private def twoForOne(apples: List[Apple.type]): List[Apple.type] = {
-    val totalSize = calculateTotalSize(apples, 2, 1)
-    List.fill(totalSize)(Apple)
-  }
+  private def twoForOne(apples: List[Fruit]): List[Fruit] = List.fill(calculateTotalSize(apples, 2, 1))(Apple)
 
-  private def threeForTwo(oranges: List[Orange.type]): List[Orange.type] = {
-    val totalSize = calculateTotalSize(oranges, 3, 2)
-    List.fill(totalSize)(Orange)
-  }
+  private def threeForTwo(oranges: List[Fruit]): List[Fruit] = List.fill(calculateTotalSize(oranges, 3, 2))(Orange)
 
-  private def calculateTotalSize(xs: List[Fruit], buy: Int, free: Int): Int = {
-    val size = xs.size
-    (size % buy) + ((size / buy) * free)
-  }
+  private def calculateTotalSize(xs: List[Fruit], buy: Int, free: Int): Int = ((s: Int) => (s % buy) + ((s / buy) * free)) (xs.size)
```

Comment:
yes, the goal was to remove braces and reduce line count. readability not my concern ;-)